### ISSUE TITLE
Bad error message and process panic when using exported library functions.

### ISSUE THREAD

**Issue by GrahamDumpleton** [k14s/ytt#687]
**What steps did you take, what happened and what did you expect:**

Create a single level library structure.

```
.
├── default.yaml
├── _ytt_lib
│   └── mylib
│       └── local.star
```

In `default.yaml` add:

```
#@ load("@ytt:library", "library")
#@ load("@ytt:struct", "struct")
#@ load("@ytt:template", "template")
#@ mylib = library.get("mylib")
#@ func = mylib.export("func")
_: #@ template.replace(func(struct.make()))
```

In `_ytt_lib/mylib/local.star` add:

```
def func(obj):
  return {"xxx": 1}
  # return {"xxx": obj.xxx}
  # xxx
end
```

Run `ytt -f .` and you get the output:

```
xxx: 1
```

Now change `_ytt_lib/mylib/local.star` to:

```
def func(obj):
  # return {"xxx": 1}
  # return {"xxx": obj.xxx}
  xxx
end
```

Run `ytt -f .` and you get the output:

```
ytt: Error:
- library.export: Exporting from library 'mylib': Evaluating starlark template:
    in <toplevel>
      default.yaml:5 | #@ func = mylib.export("func")

    reason:
     - undefined: xxx
         4 |   xxx
```

So as expected get nice error message about what the problem is:

Now change `_ytt_lib/mylib/local.star` to:

```
def func(obj):
  # return {"xxx": 1}
  return {"xxx": obj.xxx}
  # xxx
end
```

Run `ytt -f .` and you get the output:

```
panic: Expected to find compiled template: Expected to find '_ytt_lib/mylib/local.star' compiled template

goroutine 1 [running]:
github.com/vmware-tanzu/carvel-ytt/pkg/template.CompiledTemplateMultiError.buildPos({{0x0, 0x0, 0x0}, {0x166a9a8, 0xc000100f60}}, {0xc0001dcd90, 0x0, 0x0})
	github.com/vmware-tanzu/carvel-ytt/pkg/template/compiled_template_error.go:161 +0x39d
github.com/vmware-tanzu/carvel-ytt/pkg/template.CompiledTemplateMultiError.buildEvalErr({{0x0, 0x0, 0x0}, {0x166a9a8, 0xc000100f60}}, 0xc0001db650)
	github.com/vmware-tanzu/carvel-ytt/pkg/template/compiled_template_error.go:144 +0x125
github.com/vmware-tanzu/carvel-ytt/pkg/template.NewCompiledTemplateMultiError({0x1663c80, 0xc0001db650}, {0x166a9a8, 0xc000100f60})
	github.com/vmware-tanzu/carvel-ytt/pkg/template/compiled_template_error.go:57 +0x349
github.com/vmware-tanzu/carvel-ytt/pkg/template.(*CompiledTemplate).Eval(0xc000101260, 0xc0001026e0, {0x166a9a8, 0xc000100f60})
	github.com/vmware-tanzu/carvel-ytt/pkg/template/compiled_template.go:124 +0x594
github.com/vmware-tanzu/carvel-ytt/pkg/workspace.(*TemplateLoader).EvalYAML(0xc000100f60, {0xc0001026e0, 0xc0001026e0}, 0xc000102690)
	github.com/vmware-tanzu/carvel-ytt/pkg/workspace/template_loader.go:199 +0x66c
github.com/vmware-tanzu/carvel-ytt/pkg/workspace.(*LibraryExecution).eval(0xc0001f7a48, 0xc00011b080, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0})
	github.com/vmware-tanzu/carvel-ytt/pkg/workspace/library_execution.go:181 +0x90d
github.com/vmware-tanzu/carvel-ytt/pkg/workspace.(*LibraryExecution).Eval(0xc0001f7a48, 0x0, {0x0, 0x0, 0xc00000c648}, {0x0, 0xc00006b960, 0x0})
	github.com/vmware-tanzu/carvel-ytt/pkg/workspace/library_execution.go:133 +0x46
github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template.(*Options).RunWithFiles(0xc00015a840, {{0xc00006b9c0, 0x2, 0x100c00013bae0}}, {0x1670af0, 0xc00019b3e0})
	github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template/cmd.go:159 +0x550
github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template.(*Options).Run(0xc00015a840)
	github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template/cmd.go:98 +0x4b1
github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template.NewCmd.func1(0x1199fe5, {0xc00007e9a0, 0x0, 0x0})
	github.com/vmware-tanzu/carvel-ytt/pkg/cmd/template/cmd.go:63 +0x1d
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc000146600, {0xc00007e9a0, 0x0, 0x2})
	github.com/cppforlife/cobrautil@v0.0.0-20200514214827-bb86e6965d72/misc.go:45 +0x6d
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc000186a00, {0xc00007e9a0, 0x0, 0x2})
	github.com/cppforlife/cobrautil@v0.0.0-20200514214827-bb86e6965d72/misc.go:45 +0x6d
github.com/spf13/cobra.(*Command).execute(0xc000186a00, {0xc00001e1c0, 0x2, 0x2})
	github.com/spf13/cobra@v1.3.0/command.go:856 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc000186a00)
	github.com/spf13/cobra@v1.3.0/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.3.0/command.go:902
main.main()
	github.com/vmware-tanzu/carvel-ytt/cmd/ytt/ytt.go:21 +0xd6
```

What I would instead expect to see is a nice error message which includes:

```
    reason:
     - struct has no .xxx field or method
         in func
           3 |   return {"xxx": obj.xxx}
```

So there is something strange about how an error is handled for the case of failed attribute lookup in a struct object when the function is exported from a library and called.

If change `default.yaml` to:

```
#@ load("@ytt:library", "library")
#@ load("@ytt:struct", "struct")
#@ load("@ytt:template", "template")
#@ mylib = library.get("mylib")
#@ func = mylib.export("func")
_: #@ template.replace(func(struct.make(xxx=1)))
```

such that the passed struct object has the attribute, then it works fine and you get:

```
xxx: 1
```

**Anything else you would like to add:**

Nope.

**Environment:**

- ytt version (use `ytt --version`):

```
ytt version 0.40.1
```

- OS (e.g. from `/etc/os-release`):

```
% uname -a
Darwin grumpy-in-office.local 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64 x86_64
```

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

**Comment by GrahamDumpleton** [comment]
Related to issue https://github.com/vmware-tanzu/carvel-ytt/issues/156

**Comment by pivotaljohn** [comment]
@GrahamDumpleton — thank you for this thorough treatment of this bug. The additional detail will likely prove to be helpful in shaping the ultimate fix. 🙏🏻 

This issue and #156 are not just related, but point to the exact same defect.
I've highlighted the fact that details in this issue should be considered when troubleshooting.

For clarity's sake, I'll close this issue in favor of keeping the history on the older issue.

**Comment by pivotaljohn** [comment]
Released in [v0.42.0](https://github.com/vmware-tanzu/carvel-ytt/releases/tag/v0.42.0).
